pkg/scm/docker/copy.go

```diff
@@ -0,0 +1,88 @@
+package docker
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/openshift/source-to-image/pkg/api"
+	dockerpkg "github.com/openshift/source-to-image/pkg/docker"
+	"github.com/openshift/source-to-image/pkg/errors"
+	"github.com/openshift/source-to-image/pkg/scm/file"
+	"github.com/openshift/source-to-image/pkg/tar"
+	"github.com/openshift/source-to-image/pkg/util"
+)
+
+type Docker struct {
+}
+
+func (d *Docker) Download(config *api.Config) (*api.SourceInfo, error) {
+	dockerclient, err := dockerpkg.New(config.DockerConfig, config.PullAuthentication)
+	if err != nil {
+		return nil, err
+	}
+	image, sourcePath := parseDockerSource(config.Source)
+	targetDir := filepath.Join(config.WorkingDir, "upload", "sources")
+
+	if config.ForcePull {
+		if _, err = dockerclient.PullImage(image); err != nil {
+			glog.Warningf("Failed to pull %q (%v), searching for the local image ...", image, err)
+			_, err = dockerclient.CheckImage(image)
+		}
+	} else {
+		_, err = dockerclient.CheckAndPullImage(image)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Copy the files from the container
+	outReader, outWriter := io.Pipe()
+	errReader, errWriter := io.Pipe()
+	defer errReader.Close()
+	defer errWriter.Close()
+	glog.V(1).Infof("Downloading sources from %q to %q", image, targetDir)
+	extractFunc := func() error {
+		defer outReader.Close()
+		return tar.New().ExtractTarStream(targetDir, outReader)
+	}
+	opts := dockerpkg.RunContainerOptions{
+		Image:       image,
+		ScriptsURL:  config.ScriptsURL,
+		Destination: config.Destination,
+		Cmd:         []string{"tar", "cf", "-", sourcePath},
```

Comment:
this is the only thing that sucks :/ the source Docker image must have 'tar' so I can stream the directory out from it... which means you can't use `FROM scratch` but you have to `FROM busybox` ;-)